lib/WorseReflection/Core/Inference/GenericMapResolver.php

```diff
@@ -0,0 +1,64 @@
+<?php
+
+namespace Phpactor\WorseReflection\Core\Inference;
+
+use Phpactor\WorseReflection\Core\ClassName;
+use Phpactor\WorseReflection\Core\Reflector\ClassReflector;
+use Phpactor\WorseReflection\Core\TemplateMap;
+use Phpactor\WorseReflection\Core\Type;
+use Phpactor\WorseReflection\Core\Type\ClassType;
+use Phpactor\WorseReflection\Core\Type\GenericClassType;
+
+class GenericMapResolver
+{
+    private ClassReflector $reflector;
+
+    public function __construct(ClassReflector $reflector)
+    {
+        $this->reflector = $reflector;
+    }
+
+    /**
+     * @param Type[] $arguments
+     */
+    public function resolveClassTemplateMap(ClassType $topClass, ClassName $bottomClass, array $arguments = []): ?TemplateMap
+    {
+        $topReflection = $this->reflector->reflectClassLike($topClass->name());
+
+        $templateMap = $topReflection->templateMap();
+        $templateMap = $templateMap->mapArguments($arguments);
+
+        if ($topClass->name() == $bottomClass) {
+            return $templateMap;
+        }
+
+        foreach (array_merge(
+            $topReflection->docblock()->implements(),
+            $topReflection->docblock()->extends()
+        ) as $genericClass
+        ) {
+            if (!$genericClass instanceof GenericClassType) {
+                // todo: trigger a diagnostic?
```

Comment:
probably not in this PR, but would be good to have a diagnostic buffer where we can inject diagonstics for stuff like this